File: zig/src/problem0002/linked_list.zig
const std = @import("std");

pub const ListNode = struct {
    val: i32 = 0,
    next: ?*ListNode = null,
};

pub fn int_2_list(nums: []const i32) ?*ListNode {
    const length = nums.len;

    if (length == 0) {
        return null;
    }

    var head: ?*ListNode = null;

    for (0..length) |i| {
        const node = std.heap.page_allocator.create(ListNode) catch unreachable;

        node.* = .{
            .val = nums[length - i - 1],
        };
        node.next = head;
        head = node;
    }

    return head;
}

File: rust/src/problem0002/linked_list.rs
#[derive(Debug, PartialEq, Eq)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        Self { val, next: None }
    }
}

pub fn int_2_list(nums: Vec<i32>) -> Option<Box<ListNode>> {
    let mut head = None;

    for &num in nums.iter().rev() {
        let mut node = ListNode::new(num);
        node.next = head;
        head = Some(Box::new(node));
    }

    head
}
